hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/oracle/OracleUpsertQueryBuilder.java

```diff
@@ -0,0 +1,144 @@
+ * Licensed under the Hazelcast Community License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.connector.jdbc.oracle;
+
+import com.hazelcast.jet.sql.impl.connector.jdbc.AbstractQueryBuilder;
+import com.hazelcast.jet.sql.impl.connector.jdbc.JdbcTable;
+import org.apache.calcite.sql.SqlDialect;
+
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Upsert statement builder for Oracle
+ */
+public class OracleUpsertQueryBuilder extends AbstractQueryBuilder {
+
+    public OracleUpsertQueryBuilder(JdbcTable jdbcTable, SqlDialect dialect) {
+        super(jdbcTable, dialect);
+
+        StringBuilder sb = new StringBuilder();
+
+        appendMergeClause(sb);
+        sb.append(' ');
+        appendMatchedClause(sb);
+        //sb.append(" EXIT; EXCEPTION WHEN NO_DATA_FOUND THEN NULL; WHEN DUP_VAL_ON_INDEX THEN NULL; END; END LOOP;");
+
+        query = sb.toString();
+    }
+
+    void appendMergeClause(StringBuilder sb) {
+        //sb.append("LOOP BEGIN ");
+        sb.append("MERGE INTO ");
+        dialect.quoteIdentifier(sb, jdbcTable.getExternalNameList());
+        sb.append(" M USING (SELECT");
+        Iterator<String> it = jdbcTable.dbFieldNames().iterator();
+        while (it.hasNext()) {
+            String dbFieldName = it.next();
+            sb.append(" ? as ");
+            sb.append("\"" + dbFieldName + "\"");
+
+            if (it.hasNext()) {
+                sb.append(',');
+            }
+        }
+        sb.append(" FROM dual) E");
+        sb.append(" ON (");
+        appendPrimaryKeys(sb);
+        sb.append(")");
+    }
+
+    void appendMatchedClause(StringBuilder sb) {
+        sb.append("WHEN MATCHED THEN ");
+        sb.append("UPDATE ");
+        sb.append(" SET");
+        List<String> pkList = jdbcTable.getPrimaryKeyList();
+        Iterator<String> it = jdbcTable.dbFieldNames().iterator();
+        while (it.hasNext()) {
+            String dbFieldName = it.next();
+            boolean isPk = false;
+            // Oracle doesn't allow updating the values referenced in the ON clause i.e. the primary keys
+            // Skip the primary keys.
+            for (String field : pkList) {
+                if (field.equals(dbFieldName)) {
+                    isPk = true;
+                    break;
+                }
+            }
+
+            if (isPk) {
+                continue;
+            }
+
+            sb.append(" M.");
+            sb.append("\"" + dbFieldName + "\"");
+            sb.append(" = E.");
```

Comment:
WTF is M and E? :-)
Could we use some more descriptive names?

Reply:
:D yes of course, in some example they used it as aliases and it must have stuck into my mind.